### ISSUE TITLE
[Bug][DSV4][Blackwell SM12x] Indexer decode paged MQA logits broken when DeepGEMM is available: 2-state compress pages vs kernels templated for block_kv 32/64

### ISSUE THREAD

**Issue by kitch2400** [vllm-project/vllm#53635]
Summary
DeepSeek-V4 (and V4-Flash) cannot start on Blackwell client GPUs (GB10 / DGX Spark, SM121; likely also SM120) whenever DeepGEMM is importable, because the sparse attention indexer decode path unconditionally routes logits computation into DeepGEMM's paged MQA kernel, which only supports physical page geometries of 32 or 64 states. Since the #51718 KV cache layout standardization, DSV4 uses tokens_per_state = compress_ratio = 128, i.e. kv_cache_spec.num_states = 2. There is no working configuration: pristine main crashes with a clean host assert, and the commonly attempted workaround (gating metadata construction on num_states, as in open PR #53522) trades the loud assert for a silent deferred illegal memory access that surfaces far away from the real culprit.

Environment
2x NVIDIA DGX Spark GB10 (SM121, aarch64), TP=2 native distributed, no Ray. Model deepseek-ai/DeepSeek-V4-Flash-0731 (quantization_config.scale_fmt = ue8m0), dspark speculative decoding next_n=6, kv_cache_dtype fp8, max_model_len 1000000. vLLM built from source at multiple mains between Aug 22 and Aug 24 (reproduced on 0ab4d638a and 6648eb11), torch 2.13.0+cu130, vendored DeepGEMM built from DEEPGEMM_REF=a6b593d, FlashInfer ff22228d.

Failure mode A, pristine main
During compile_or_warm_up_model, on the first decode-shaped forward, get_paged_mqa_logits_metadata asserts:

Assertion error (csrc/apis/attention.hpp:247): block_kv == 32 or block_kv == 64

because self.kv_cache_spec.num_states is 2 for compress-128 pages. This is unconditional in current main: the call site at the end of the decode metadata builder checks nothing but platform.is_cuda() and has_deep_gemm().

Failure mode B, with a num_states gate (open PR #53522)
Adding the obvious gate (only build metadata when num_states is 32 or 64) removes the assert but leaves schedule_metadata as an unfilled zero buffer, and the decode path still launches smxx_fp8_fp4_paged_mqa_logits unconditionally on CUDA. The kernel consumes the garbage schedule and corrupts memory asynchronously. The sticky CUDA_ERROR_ILLEGAL_ADDRESS then surfaces at an arbitrary later point - in our case persistent_topk's occupancy query (topk.cu:106) during graph capture, and with CUDA_LAUNCH_BLOCKING=1 it resolves honestly to smxx_fp8_fp4_paged_mqa_logits.hpp:202. This cost us the better part of a day because the visible error pointed at completely unrelated kernels.

Root cause
block_kv is a compile-time template parameter of the paged MQA kernel (baked into the JIT instantiation), the metadata scheduler works in units of physical KV blocks, and DSV4's compressed geometry (128 tokens per state, 2 states per page) simply does not fit either supported tile. The pre-standardization tree worked because storage_block_size (64) flowed into the helper natively.

Notes
- Warmup executes decode-shaped forwards regardless of --enforce-eager, so eager mode is not an escape hatch; the same kernel IMAs on the first real decode request anyway.
- The archived megapatch #41834 solved this for SM12x with a dedicated short-row decode top-k subsystem (fused fp8_fp4_paged_mqa_topk_indices kernel plus width-based fallbacks, csrc/libtorch_stable changes). That subsystem has not been sliced upstream; open PR #53522 addresses only the metadata assert and is insufficient on its own for the reasons above.
- Related: #52732 (unguarded DeepGEMM assumptions on sm_121), #47988 (E8M0 block scales in CUTLASS/Triton linear kernels, also required to boot this checkpoint).

Question for maintainers: what is the preferred direction here? Options we see are (a) teaching the arch-12 paged kernel to handle smaller block_kv geometries, (b) slicing the megapatch's short-row decode top-k path as the official SM12x indexer decode route, or (c) a non-DeepGEMM fallback for 2-state pages. Happy to test patches on real hardware; we have a fully instrumented reproduction including per-shape operand dumps from live captures.

**Comment by kitch2400** [comment]
Gentle bump, plus some new info that narrows this further.

maci0 confirmed in #53680 that the a6b593d DeepGEMM build runs the SM12x fp8 einsum path cleanly on GB10 - we're serving DSV4-Flash right now on a dual Spark cluster pinned there, and the indexer decode geometry described above is the only thing between us and a fully working stack. Everything else we fought through last week (einsum recipe/shape handling on arch 12, E8M0 scales in CUTLASS) is solved or has an open PR.

Still reproduces on current main as of today (73bd7c83c) - the decode metadata builder still passes num_states=2 straight into the helper.

So of everything blocking DSV4-Flash on Blackwell client GPUs, this is the last one standing. Would really appreciate knowing which direction maintainers would prefer: teaching the arch-12 paged kernel to handle smaller block_kv, slicing the megapatch short-row top-k subsystem as the official SM12x route, or something else entirely. We've got hardware warm and can test patches same-day.

**Comment by ivanusto** [comment]
Two datapoints from another 2x GB10 pair (sm_121a, TP=2 over RoCE), one confirming and one that does not fit, plus a weightless repro for the assert.

## 1. The assert is real on the current upstream pin, and reproduces with no weights

Stock `vllm/vllm-openai:nightly-aarch64` (`v0.26.1rc1.dev1219+g46638857f`, vendored DeepGEMM at the `8b1392b` cmake pin). Calling the helper directly, no model, no server:

```python
import torch
from vllm.utils.deep_gemm import get_paged_mqa_logits_metadata
num_sms = torch.cuda.get_device_properties(0).multi_processor_count   # 48 on GB10
seq_lens = torch.full((4, 1), 64, dtype=torch.int32, device="cuda")
for ns in (2, 16, 32, 64, 256):
    try:
        m = get_paged_mqa_logits_metadata(seq_lens, ns, num_sms)
        print(ns, "OK", tuple(m.shape))
    except Exception as e:
        print(ns, "RAISE", e)
```

```
num_states=   2  RAISE -> Assertion error (csrc/apis/attention.hpp:262): block_kv == 32 or block_kv == 64
num_states=  16  RAISE -> Assertion error (csrc/apis/attention.hpp:262): block_kv == 32 or block_kv == 64
num_states=  32  OK    -> (49, 2) torch.int32
num_states=  64  OK    -> (49, 2) torch.int32
num_states= 256  RAISE -> Assertion error (csrc/apis/attention.hpp:262): block_kv == 32 or block_kv == 64
```

So the geometry constraint you describe holds on `8b1392b` as well as on the `a6b593d` you built against (the assert moved from line 247 to 262 between the two, same condition). This takes ~10 seconds in a `--gpus all` container and needs no checkpoint, which may be useful for bisecting patches.

## 2. But the full model boots and serves on this pair with DeepGEMM available

Same image, unmodified, no DeepGEMM environment variables set at all (`has_deep_gemm()` returns True, the DeepGEMM 984-kernel warmup runs, and `DeepGemmFp8BlockScaledMMKernel` is selected for `Fp8LinearMethod`). DeepSeek-V4-Flash-0731, NVFP4 conversion, `scale_fmt = ue8m0`, `--tensor-parallel-size 2 --nnodes 2 --kv-cache-dtype fp8_ds_mla --max-model-len 8192`, no speculative decoding.

It completes weight load, `compile_or_warm_up_model`, PIECEWISE 51/51 and FULL 2/2 graph capture, and serves. Three 400-token greedy generations come back coherent at a median 26.76 tok/s. **No `attention.hpp` assert, no deferred illegal access, nothing in the log.**

## 3. Why that does not fit, stated as an open question rather than a claim

The checkpoint here has an alternating compress ratio rather than a uniform 128:

```
compress_ratios = [0, 0, 4, 128, 4, 128, 4, 128, ... 4, 128, 4, 0, 0, 0]
index_head_dim = 128, index_n_heads = 64, index_topk = 512, dspark_block_size = 5
```

By my reading of the current tree that should still hit your assert:

- `DeepseekV4IndexerBackend.get_supported_kernel_block_sizes()` returns `[256]`.
- `AttentionSpec.num_states` is `kernel_block_size // tokens_per_state`, so a ratio-128 layer gives `256 // 128 = 2`.
- `MLAAttentionSpec.merge` asserts `len(tokens_per_state_set) == 1`, so layers with different ratios cannot be folded into one cache group and the ratio-128 layers should keep their own spec.

By that chain the ratio-128 group should reach `indexer.py:1040` with `num_states = 2` and assert, exactly as you describe. It does not, and I have not localized why. I am not claiming the bug is checkpoint-dependent; I am reporting a same-hardware configuration where the documented path does not fire, in case it helps bound the trigger.

If it is useful I can instrument `kv_cache_spec.tokens_per_state` and `num_states` per cache group on this pair and post what the DSV4 indexer group actually resolves to, or run any specific shape or patch you want checked. A negative case with a known-good boot may be easier to bisect against than a failing one.

## 4. On #53522

Your point that gating metadata construction alone converts a loud assert into a deferred illegal access matches what the call site looks like: `schedule_metadata` stays as the unfilled `scheduler_metadata_buffer` and the decode kernel is launched regardless. Worth noting on that PR directly, since it has no maintainer review yet and the gate reads as sufficient without your analysis.

**Comment by ivanusto** [comment]
Following up on my open question above. I traced it, and it turns out to need no model load at all. Posting the chain because it does not line up with the `tokens_per_state = 128` premise, and you are better placed than me to say where the difference is.

## On this tree the indexer's `tokens_per_state` is structurally 4, never 128

`vllm/models/deepseek_v4/attention.py`:

```python
# :212
self.compress_ratio = max(1, config.compress_ratios[layer_id])

# :276
self.indexer = None
if self.compress_ratio == 4:
    # Only C4A uses sparse attention and hence has indexer.
    self.indexer = DeepseekV4Indexer(..., compress_ratio=self.compress_ratio, ...)

# :844, inside DeepseekV4Indexer.__init__
self.k_cache = DeepseekV4IndexerCache(..., compress_ratio=self.compress_ratio)

# :757, DeepseekV4IndexerCache.get_kv_cache_spec
return MLAAttentionSpec(block_size=self.cache_config.block_size,
                        tokens_per_state=self.compress_ratio, ...)
```

The compress-128 layers never construct an indexer. They go through `self.compressor`
(`DeepseekCompressor`, `attention.py:344`), which does not reach
`get_paged_mqa_logits_metadata`. So the only spec that feeds `indexer.py:1043` carries
`tokens_per_state = 4`, and with `DeepseekV4IndexerBackend.get_supported_kernel_block_sizes()`
returning `[256]`, `AttentionSpec.num_states` is `256 // 4 = 64`, which is one of the two
values the kernel accepts.

Replaying that arithmetic with vLLM's own `MLAAttentionSpec`, no weights and no GPU needed:

```
compress_ratios = [0, 0, 4, 128, 4, 128, ... 4, 128, 4, 0, 0, 0]   (num_hidden_layers = 43)

layers with an indexer (ratio == 4): 21
layers with an MLA spec (ratio > 1): 41, ratios {4, 128}

  indexer group (C4A only)      tokens_per_state=  4   num_states= 64   -> PASS
  MLA group ratio=4             tokens_per_state=  4   num_states= 64
  MLA group ratio=128           tokens_per_state=128   num_states=  2   -> would assert
```

The ratio-128 MLA group is the only place a `num_states = 2` appears, and that group is not
the one whose builder calls the paged MQA helper. That is why the boot here is clean while
the direct kernel probe in my previous comment still asserts on 2.

## Two things that make me unsure about the premise rather than about your observation

1. **The official checkpoint is not uniform 128 either.** Reading
   `config.json` from a local snapshot of `deepseek-ai/DeepSeek-V4-Flash-0731`
   (`9e165c30e2704aec5d9d593cce3eebd58bbef1cb`) gives the same alternating list,
   `[0, 0, 4, 128, 4, 128, ... 4, 0, 0, 0]`, identical to the NVFP4 conversion I run.
   So there should be 21 ratio-4 layers on your checkpoint too.
2. **The C4A gating is not new.** It is present at `6648eb11`, one of the two commits you
   say you reproduced on:
   `gh api repos/vllm-project/vllm/contents/vllm/models/deepseek_v4/attention.py?ref=6648eb11`
   shows `if self.compress_ratio == 4:` with the same `Only C4A uses sparse attention`
   comment at line 277.

So I cannot get from "DSV4 uses tokens_per_state = compress_ratio = 128" to the indexer
builder on the trees I can read. You have an instrumented reproduction with a real IMA and
I do not, so I am not claiming your root cause is wrong. But if the 2 you observe is coming
from the ratio-128 MLA cache group rather than the indexer group, the fix surface changes:
it would be about which spec the indexer builder is bound to, not about teaching the arch-12
paged kernel smaller `block_kv` geometries.

Two questions you can answer from your existing capture, and which would settle it:

- At the point of the assert, what does `self.kv_cache_spec.tokens_per_state` print, 4 or 128?
- Which cache group's metadata builder is on the stack, the `DEEPSEEK_V4_INDEXER` one or an MLA one?

Happy to run anything else here. The spec replay above takes a few seconds in a CPU-only
container, so it is easy to re-run against whatever config you want compared.

**Comment by ivanusto** [comment]
One update on the DeepGEMM side of this, since the pin moved after my last comment.

#57218 (merged 09-16) moved `deepgemm.cmake` to `vllm-project/DeepGEMM@a6bbb800`, and that range rewrites `sm120_fp8_paged_mqa_logits` and `sm120_fp8_mqa_logits` among others. That could have changed what the paged MQA metadata call accepts, so I re-ran it on a GB10 (sm_121) against both pins, same image, weight free.

`get_paged_mqa_logits_metadata`, states 2 / 16 / 32 / 64:

| pin | 2 | 16 | 32 | 64 |
|---|---|---|---|---|
| `ad1f1726` | assert | assert | ok | ok |
| `a6bbb80` | assert | assert | ok | ok |

Identical, only the assert's line number moved (408 to 409). So the pin bump does not change the premise of this issue: the kernel still takes 32 and 64 only, which is what my 09-02 trace landed on, where the indexer spec yields `num_states = 256 // 4 = 64` and therefore stays inside the accepted set on this tree.

`test_deepgemm_attention.py` is also unchanged across the two pins here: 4 passed, 1 failed, the failure being the known `test_deepgemm_fp8_mqa_logits[True]` `clean_logits=True` assert, which the serving path does not hit because it passes `clean_logits=False`.

Worth knowing for anyone re-testing this: those sm120 kernel changes have no CI covering them. `deepgemm.cmake` only triggers the H100 DeepGEMM job, both DGX Spark jobs are `optional: true` and list no DeepGEMM files, and the fork's own workflow builds 9.0+PTX and below, so the fork's `tests/test_sm120_*.py` never runs anywhere.
